runtime/compiler/env/j9methodServer.cpp

```diff
@@ -1002,80 +994,15 @@ TR_ResolvedJ9JITServerMethod::getResolvedVirtualMethod(TR::Compilation * comp, T
    }
 
 char *
-TR_ResolvedJ9JITServerMethod::getRemoteROMString(int32_t &len, void *basePtr, std::initializer_list<size_t> offsets)
-   {
-   TR_ASSERT(offsets.size() <= 2, "Number of offsets is greater than 2");
-
-   auto offsetFirst = *offsets.begin();
-   auto offsetSecond = (offsets.size() == 2) ? *(offsets.begin() + 1) : 0;
-
-   TR_ASSERT(offsetFirst < (1 << 16) && offsetSecond < (1 << 16), "Offsets are larger than 16 bits");
-
-   // create a key for hashing into a table of strings
-   TR_RemoteROMStringKey key;
-   uint32_t offsetKey = (offsetFirst << 16) + offsetSecond;
-   key._basePtr = basePtr;
-   key._offsets = offsetKey;
-   
-   std::string *cachedStr = NULL;
-   bool isCached = false;
-   TR::CompilationInfoPerThread *threadCompInfo = _fe->_compInfoPT;
+TR_ResolvedJ9JITServerMethod::getROMString(int32_t &len, void *basePtr, std::initializer_list<size_t> offsets)
    {
-      OMR::CriticalSection getRemoteROMClass(threadCompInfo->getClientData()->getROMMapMonitor()); 
-      auto &stringsCache = getJ9ClassInfo(threadCompInfo, _ramClass)._remoteROMStringsCache;
-      auto gotStr = stringsCache.find(key);
-      if (gotStr != stringsCache.end())
-         {
-         cachedStr = &(gotStr->second);
-         isCached = true;
-         }
+   return getJ9ClassInfo(_fe->_compInfoPT, _ramClass).getROMString(len, basePtr, offsets);
```

Comment:
```
ClientSessionData::ClassInfo &
getJ9ClassInfo(TR::CompilationInfoPerThread *threadCompInfo, J9Class *clazz)
   {
   // This function assumes that you are inside of _romMapMonitor
   // Do not use it otherwise
```
We nee to acquire `_romMapMonitor` like the old function did, buy if we need to send messages we are in trouble.

Reply:
True. Now that I think about it, the original implementation should be acquiring the monitor in `getROMString`  as well, even when we know that remote call is not needed. Because if the class gets unloaded the ROM class on the server gets deleted as well.

Reply:
Right. I think the solution to this is to make `getRemoteROMString` and `getROMString` methods of `ClientSesssionData` instead of `ClientSessionData::ClassInfo`. Then we could acquire monitor from inside `ClientSessionData`.

Now that I think about it, the part of `getROMString` where we dereference a string accessible from the server should also be guarded with a monitor. Since those strings are part of the ROM class and could be invalidated by a class unload event.